### PR TITLE
Quiz v1.3: hint and difficult command

### LINKED ISSUE #70: As a user I can see hints for a question
so that I can answer the question when I have uncertainty

### LINKED ISSUE #72: As a user I can label question as difficult
so that I can review them more frequently

### LINKED ISSUE #128: Continue quiz by typing enter when answer wrongly twice
Currently the user can only type the correct answer to continue doing the test while they answer wrongly twice.

### PR DESCRIPTION
**Pull request by halcyoneee**
**New**:
- Implemented `\difficult` quiz command fully that allows you to label the current card as difficult or not , now it will save it to `savedata.csv` (closes #72).
- Implemented `\hint` quiz command that allows you to view the current card's hint (closes #70).

**Documentation**:
- Updated DG `LogicClassDiagram`, `UiClassDiagram`, `QuizModelClassDiagram` and `QuizAnswerCommandSequenceDiagram`
- Added DG `hint` and `status` section.
- Updated quiz alternative choices and explanation.

**Bug fixes & testing**
- Updated Ui display (closes #128)

### CODE REVIEW

File: docs/team/halcyoneee.adoc [hunk 1]
```diff
@@ -28,7 +29,10 @@ commands between the two modes, quiz mode commands start with a backslash `\`, w
 commands are not.
 
 ** *Highlights*: Pull request https://github.com/CS2103-AY1819S2-W14-1/main/pull/32[#32],
-https://github.com/CS2103-AY1819S2-W14-1/main/pull/43[#42]
+https://github.com/CS2103-AY1819S2-W14-1/main/pull/43[#42],
+https://github.com/CS2103-AY1819S2-W14-1/main/pull/96[#96],
+https://github.com/CS2103-AY1819S2-W14-1/main/pull/108[#108]
```
Comment (DigiPie) [inline docs/team/halcyoneee.adoc]: Missing a comma at end of line.

```suggestion
https://github.com/CS2103-AY1819S2-W14-1/main/pull/108[#108],
```

File: src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Displays hint of the current card
+ */
+public class QuizHintCommand extends QuizCommand {
+    public static final String COMMAND_WORD = "\\hint";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Force quits this session, "
+        + "only progress of attempted questions will still be saved.\n"
+        + "Example: " + COMMAND_WORD + "\n";
```
Comment (DigiPie) [inline src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java]: Incorrect `MESSAGE_USAGE`. Please rectify.

File: src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    public static final String MESSAGE_SUCCESS = "Hints: %1$s\n";
+    public static final String MESSAGE_EMPTY = "No hint for this card\n";
```
Comment (DigiPie) [inline src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java]:
```suggestion
    public static final String MESSAGE_EMPTY = "No hint for this card.\n";
```

File: src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Displays hint of the current card
+ */
+public class QuizHintCommand extends QuizCommand {
+    public static final String COMMAND_WORD = "\\hint";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Force quits this session, "
+        + "only progress of attempted questions will still be saved.\n"
+        + "Example: " + COMMAND_WORD + "\n";
+
+    public static final String MESSAGE_SUCCESS = "Hints: %1$s\n";
```
Comment (DigiPie) [inline src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java]: Hint or hints? On line 14 it was stated that the class 'displays hint of the current card'.

File: src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        requireNonNull(model);
+        QuizModel quizModel = requireQuizModel(model);
+
+        List<String> hints = quizModel.getOpt();
+
+        if (hints.isEmpty()) {
+            throw new CommandException(MESSAGE_EMPTY);
+        }
+
+        return new CommandResult(String.format(MESSAGE_SUCCESS, quizModel.getOpt().toString()));
```
Comment (DigiPie) [inline src/main/java/seedu/address/logic/commands/quiz/QuizHintCommand.java]: Redundant call to `quizModel.getOpt()`. Make use of earlier call and result stored in `List<String> hints`.

```suggestion
        return new CommandResult(String.format(MESSAGE_SUCCESS, hints.toString()));
```

File: src/main/java/seedu/address/model/modelmanager/QuizModel.java [hunk 6]
```diff
@@ -93,6 +84,21 @@
...
+     */
+    String getQuestionHeader();
+
+    /**
+     * Returns the answer header of a {@code QuizCard}
+     */
+    String getAnswerHeader();
+
+    /**
+     * Returns list of optional of the original {@code QuizCard} with index.
```
Comment (DigiPie) [inline src/main/java/seedu/address/model/modelmanager/QuizModel.java]: What is the `index` you are referring to in 'with index'?

File: src/main/java/seedu/address/model/quiz/Quiz.java [hunk 7]
```diff
@@ -63,12 +62,12 @@ public Quiz(List<QuizCard> session, QuizMode mode) {
         case REVIEW:
             generateReview();
             break;
+        case DIFFICULT:
+            generatePreview();
+            break;
         default:
             break;
```
Comment (DigiPie) [inline src/main/java/seedu/address/model/quiz/Quiz.java]: There's no purpose in adding a `break;` at the end of a `switch` block.

```suggestion
```

File: src/main/java/seedu/address/model/quiz/Quiz.java [hunk 8]
```diff
@@ -63,12 +62,12 @@ public Quiz(List<QuizCard> session, QuizMode mode) {
         case REVIEW:
             generateReview();
             break;
+        case DIFFICULT:
+            generatePreview();
+            break;
         default:
```
Comment (DigiPie) [inline src/main/java/seedu/address/model/quiz/Quiz.java]: There's no purpose in adding a `default:` if it is not going to do anything.

```suggestion
```

File: src/main/java/seedu/address/model/quiz/Quiz.java [hunk 9]
```diff
@@ -172,6 +171,40 @@ public boolean toggleIsCardDifficult(int index) {
...
+
+        return original.getQuestionHeader();
+    }
+
+    public List<String> getOpt() {
+        assert currentCardIndex != -1;
+
+        QuizCard original = currentSession.get(currentQuizCard.getIndex());
+
+        return original.getOpt();
```
Comment (DigiPie) [inline src/main/java/seedu/address/model/quiz/Quiz.java]: Should inline variable.

```suggestion
        return currentSession.get(currentQuizCard.getIndex()).getOpt();
```

File: src/main/java/seedu/address/model/quiz/Quiz.java (hunk 7 above)
Comment (halcyoneee) [reply src/main/java/seedu/address/model/quiz/Quiz.java]: For switch block, codacy will flag it as fallthrough and it is not allowed

File: src/main/java/seedu/address/model/quiz/Quiz.java (hunk 8 above)
Comment (halcyoneee) [reply src/main/java/seedu/address/model/quiz/Quiz.java]: Just fulfilling codacy requirement, it must have a default case

File: src/main/java/seedu/address/model/quiz/Quiz.java (hunk 9 above)
Comment (halcyoneee) [reply src/main/java/seedu/address/model/quiz/Quiz.java]: Following Law of Demeter, hence I had a local variable instead of using inline variable

**Comment by halcyoneee** [comment]
> Good job improving coverage and refactoring your code for better clarity and efficiency. However, there are some minor issues with your PR. Do take a look and rectify.

Thanks for pointing out the issues, overlooked it while I was rushing it at night.

**Pull request merged**